File: Sources/NockerlDesign/Components/NockerlDatePickers.swift
// NockerlDatePickers: the shared date/time picker CONTRACT on SwiftUI.
//
// Wraps the platform `DatePicker` (law §9, native interaction; the CONTRACT unifies):
//   - `weekStartsOn` (0=Sunday…6=Saturday): DEFAULTS to the device locale; the explicit
//     override IS honored here. SwiftUI derives its calendar from the environment, so a
//     custom `Calendar.firstWeekday` flows in (Apple's firstWeekday is 1-based).
//   - `entryMode` (`default | dial | wheel | text`): defaults to the platform idiom
//     (graphical calendar / wheel where native). `dial` is not an Apple idiom and falls
//     back to the default (documented, per the contract's fallback rule).
//   - `minuteStep` / `min` / `max`: the shared web API. SwiftUI's DatePicker has no
//     native minute stepping. Selections SNAP to the step on change (the documented
//     clamping behavior, same pure rule as Compose); `min`/`max` clamp via the native
//     range parameter.

import SwiftUI

/// The shared entry-mode vocabulary.
public enum NockerlPickerEntryMode: Equatable {
    /// The platform idiom (graphical calendar on macOS/iOS).
    case `default`
    /// Not an Apple idiom; falls back to `.default` (documented).
    case dial
    /// The wheel (iOS native; on macOS falls back to `.default`).
    case wheel
    /// Keyboard/field entry.
    case text
}

/// Snap a minute value to the nearest step (the contract's documented clamping;
/// identical semantics to Compose's `nockerlSnapMinutes`). `step <= 1` is a
/// no-op; 60 wraps to 0.
public func nockerlSnapMinutes(_ minute: Int, step: Int) -> Int {
    let clamped = min(max(minute, 0), 59)
    guard step > 1 else { return clamped }
    let snapped = (Int((Double(clamped) / Double(step)).rounded()) * step) % 60
    return min(max(snapped, 0), 59)
}

/// A date (or time) picker under the shared contract.
///
/// Unavailable on tvOS/watchOS: SwiftUI ships no comparable `DatePicker` there
/// within this package's platform floors.
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public struct NockerlDatePicker: View {
    private let label: String
    private let selection: Binding<Date>
    private let displayedComponents: DatePickerComponents
    private let weekStartsOn: Int?
    private let entryMode: NockerlPickerEntryMode
    private let minuteStep: Int
    private let bounds: ClosedRange<Date>?

    /// Create a picker.
    /// - Parameters:
    ///   - label: the REQUIRED persistent label (law §14).
    ///   - selection: the bound date.
    ///   - displayedComponents: `.date`, `.hourAndMinute`, or both.
    ///   - weekStartsOn: 0=Sunday…6=Saturday; `nil` follows the device locale
    ///     (the contract default).
    ///   - entryMode: shared vocabulary; unsupported values fall back (KDoc'd).
    ///   - minuteStep: snap increment for time selection (1 = free).
    ///   - min: earliest selectable instant (inclusive).
    ///   - max: latest selectable instant (inclusive).
    public init(
        _ label: String,
        selection: Binding<Date>,
        displayedComponents: DatePickerComponents = [.date],
        weekStartsOn: Int? = nil,
        entryMode: NockerlPickerEntryMode = .default,
        minuteStep: Int = 1,
        min: Date? = nil,
        max: Date? = nil
    ) {
        self.label = label
        self.displayedComponents = displayedComponents
        self.weekStartsOn = weekStartsOn
        self.entryMode = entryMode
        self.minuteStep = minuteStep
        self.bounds = Self.range(min: min, max: max)
        // The snapping binding: writes snap the minute component to the step
        // (the documented clamping; no native stepping exists).
        if minuteStep > 1 {
            self.selection = Binding(
                get: { selection.wrappedValue },
                set: { newValue in
                    selection.wrappedValue = Self.snappingMinutes(of: newValue, step: minuteStep)
                }
            )
        } else {
            self.selection = selection
        }
    }

    public var body: some View {
        PickerBody(
            label: label,
            selection: selection,
            displayedComponents: displayedComponents,
            weekStartsOn: weekStartsOn,
            entryMode: entryMode,
            bounds: bounds
        )
    }

    /// Environment-reading render body (scheme-correct brand tint).
    private struct PickerBody: View {
        let label: String
        let selection: Binding<Date>
        let displayedComponents: DatePickerComponents
        let weekStartsOn: Int?
        let entryMode: NockerlPickerEntryMode
        let bounds: ClosedRange<Date>?

        @Environment(\.colorScheme) private var colorScheme

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            let base = styled(rawPicker).tint(palette.accentPrimary)

            // weekStartsOn: honored via the environment calendar (Apple
            // firstWeekday is 1-based Sunday=1; the contract is 0-based).
            if let weekStartsOn {
                base.environment(\.calendar, NockerlDatePicker.calendar(firstWeekday: weekStartsOn))
            } else {
                base
            }
        }

        @ViewBuilder
        private var rawPicker: some View {
            if let bounds {
                DatePicker(label, selection: selection, in: bounds, displayedComponents: displayedComponents)
            } else {
                DatePicker(label, selection: selection, displayedComponents: displayedComponents)
            }
        }

        /// Map the shared entry-mode vocabulary onto the platform styles.
        /// `.dial` is not an Apple idiom and `.wheel` is iOS-only, so the
        /// documented fallbacks land on the platform default.
        @ViewBuilder
        private func styled(_ picker: some View) -> some View {
            #if os(iOS)
            switch entryMode {
            case .wheel: picker.datePickerStyle(.wheel)
            case .text: picker.datePickerStyle(.compact)
            case .default, .dial: picker.datePickerStyle(.graphical)
            }
            #elseif os(macOS)
            switch entryMode {
            case .text: picker.datePickerStyle(.field)
            case .default, .dial, .wheel: picker.datePickerStyle(.graphical)
            }
            #else
            picker
            #endif
        }
    }

    /// Build the environment calendar for an explicit week start (0-based → 1-based).
    static func calendar(firstWeekday zeroBased: Int) -> Calendar {
        var calendar = Calendar.autoupdatingCurrent
        calendar.firstWeekday = (zeroBased.clamped(to: 0...6)) + 1
        return calendar
    }

    /// Snap a date's minute component to the step (shared rule with Compose).
    static func snappingMinutes(of date: Date, step: Int) -> Date {
        let calendar = Calendar.autoupdatingCurrent
        let minute = calendar.component(.minute, from: date)
        let snapped = nockerlSnapMinutes(minute, step: step)
        guard snapped != minute else { return date }
        return calendar.date(bySetting: .minute, value: snapped, of: date) ?? date
    }

    private static func range(min: Date?, max: Date?) -> ClosedRange<Date>? {
        switch (min, max) {
        case let (lower?, upper?): return lower...upper
        case let (lower?, nil): return lower...Date.distantFuture
        case let (nil, upper?): return Date.distantPast...upper
        case (nil, nil): return nil
        }
    }
}

/// The **range MODE** of the date-picker family: a start→end
/// date range on the ONE contract, so a standalone range picker is unnecessary.
///
/// SwiftUI ships NO graphical contiguous-range picker, so honor-the-platform
/// (law §9) renders the range as two coordinated ``NockerlDatePicker``s: the
/// end is floored at the start and the start capped at the end, so the pair can
/// never invert. This is the Apple-idiomatic range affordance; the Compose half
/// uses M3's `DateRangePicker`. Same contract, platform-appropriate interaction.
///
/// Unavailable on tvOS/watchOS (as the single picker is).
@available(tvOS, unavailable)
@available(watchOS, unavailable)
public struct NockerlDateRangePicker: View {
    private let startLabel: String
    private let endLabel: String
    @Binding private var start: Date
    @Binding private var end: Date
    private let min: Date?
    private let max: Date?

    /// Create a range picker.
    /// - Parameters:
    ///   - startLabel: the start field's persistent label (default "Start").
    ///   - endLabel: the end field's persistent label (default "End").
    ///   - start: the bound range start.
    ///   - end: the bound range end.
    ///   - min: earliest selectable instant (inclusive).
    ///   - max: latest selectable instant (inclusive).
    public init(
        startLabel: String = "Start",
        endLabel: String = "End",
        start: Binding<Date>,
        end: Binding<Date>,
        min: Date? = nil,
        max: Date? = nil
    ) {
        self.startLabel = startLabel
        self.endLabel = endLabel
        self._start = start
        self._end = end
        self.min = min
        self.max = max
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: NockerlSpace.space4) {
            // Start: capped at the current end so it can never exceed it.
            NockerlDatePicker(startLabel, selection: $start, min: min, max: end)
            // End: floored at the current start so the range can never invert.
            NockerlDatePicker(endLabel, selection: $end, min: start, max: max)
        }
    }
}

private extension Int {
    /// Clamp into a closed range.
    func clamped(to range: ClosedRange<Int>) -> Int {
        Swift.min(Swift.max(self, range.lowerBound), range.upperBound)
    }
}

File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlDatePickers.kt
package com.dizyx.nockerl.design.components

import androidx.compose.material3.DatePicker
import androidx.compose.material3.DateRangePicker
import androidx.compose.material3.DisplayMode
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.SelectableDates
import androidx.compose.material3.TimeInput
import androidx.compose.material3.TimePicker
import androidx.compose.material3.rememberDatePickerState
import androidx.compose.material3.rememberDateRangePickerState
import androidx.compose.material3.rememberTimePickerState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Modifier

/**
 * The shared date/time picker CONTRACT on Compose,
 * wrapping the platform Material 3 pickers (per law §9, the interaction stays
 * native; the CONTRACT is what unifies):
 *
 * - **`weekStartsOn`** (0=Sunday … 6=Saturday): the DEFAULT is the device
 *   locale on every platform. M3's calendar derives its first day from the
 *   locale and exposes no per-day override. Under this contract, the unsupported
 *   explicit override FALLS BACK to the platform default (and this KDoc says
 *   so; the props table documents it). The prop exists so call sites are
 *   source-compatible across platforms.
 * - **`entryMode`** (`DEFAULT | DIAL | WHEEL | TEXT`): defaults to the platform
 *   idiom, namely M3 calendar (date) / clock dial (time). `TEXT` maps to M3 input
 *   modes; `WHEEL` is not an Android idiom and falls back to the default.
 * - **`minuteStep` / `minMillis` / `maxMillis`**: the web API is the shared
 *   contract. M3 has no native minute stepping. The selection SNAPS to the
 *   nearest step on read ([nockerlSnapMinutes], documented clamping); date
 *   bounds clamp via [SelectableDates].
 */
enum class NockerlPickerEntryMode {
    /** The platform idiom (M3 calendar / clock dial). */
    DEFAULT,

    /** The M3 clock dial (time; dates fall back to DEFAULT). */
    DIAL,

    /** Not an Android idiom: falls back to DEFAULT (documented). */
    WHEEL,

    /** Keyboard entry (M3 input modes). */
    TEXT,
}

/**
 * Date picker under the shared contract. Selection reports as UTC midnight
 * millis (the M3 convention).
 *
 * @param selectedDateMillis current selection (UTC millis), or `null`.
 * @param onDateSelected fires with the (bounds-clamped) selection.
 * @param modifier outer modifier.
 * @param weekStartsOn override in the 0 to 6 range (UNSUPPORTED on M3, locale-driven);
 *   documented fallback.
 * @param entryMode [NockerlPickerEntryMode.TEXT] = M3 input mode; everything
 *   else = the calendar.
 * @param minMillis earliest selectable date (UTC millis), inclusive.
 * @param maxMillis latest selectable date (UTC millis), inclusive.
 */
@Suppress("UNUSED_PARAMETER")
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NockerlDatePicker(
    selectedDateMillis: Long?,
    onDateSelected: (Long) -> Unit,
    modifier: Modifier = Modifier,
    weekStartsOn: Int? = null,
    entryMode: NockerlPickerEntryMode = NockerlPickerEntryMode.DEFAULT,
    minMillis: Long? = null,
    maxMillis: Long? = null,
) {
    val state =
        rememberDatePickerState(
            initialSelectedDateMillis = selectedDateMillis,
            initialDisplayMode =
                if (entryMode == NockerlPickerEntryMode.TEXT) DisplayMode.Input else DisplayMode.Picker,
            selectableDates = nockerlSelectableDates(minMillis, maxMillis),
        )
    LaunchedEffect(state) {
        snapshotFlow { state.selectedDateMillis }
            .collect { millis ->
                if (millis != null) {
                    onDateSelected(nockerlClampMillis(millis, minMillis, maxMillis))
                }
            }
    }
    DatePicker(state = state, modifier = modifier)
}

/**
 * The **range MODE** of the date-picker family: a start→end
 * date range on the ONE contract, so a standalone range picker is unnecessary.
 * On Compose the platform reality is M3's separate [DateRangePicker] (its state
 * differs from the single picker's), presented here as the family's range mode,
 * NOT a duplicate component (law §9/§12: honor the platform, unify the contract).
 *
 * Both ends report as UTC-midnight millis (the M3 convention); either may be
 * `null` mid-selection. Bounds clamp via the same [SelectableDates] +
 * [nockerlClampMillis] contract as the single picker.
 *
 * @param startDateMillis current range start (UTC millis), or `null`.
 * @param endDateMillis current range end (UTC millis), or `null`.
 * @param onRangeSelected fires with the (bounds-clamped) start/end, either nullable.
 * @param modifier outer modifier.
 * @param entryMode [NockerlPickerEntryMode.TEXT] = M3 input mode; else the calendar.
 * @param minMillis earliest selectable date (UTC millis), inclusive.
 * @param maxMillis latest selectable date (UTC millis), inclusive.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NockerlDateRangePicker(
    startDateMillis: Long?,
    endDateMillis: Long?,
    onRangeSelected: (start: Long?, end: Long?) -> Unit,
    modifier: Modifier = Modifier,
    entryMode: NockerlPickerEntryMode = NockerlPickerEntryMode.DEFAULT,
    minMillis: Long? = null,
    maxMillis: Long? = null,
) {
    val state =
        rememberDateRangePickerState(
            initialSelectedStartDateMillis = startDateMillis,
            initialSelectedEndDateMillis = endDateMillis,
            initialDisplayMode =
                if (entryMode == NockerlPickerEntryMode.TEXT) DisplayMode.Input else DisplayMode.Picker,
            selectableDates = nockerlSelectableDates(minMillis, maxMillis),
        )
    LaunchedEffect(state) {
        snapshotFlow { state.selectedStartDateMillis to state.selectedEndDateMillis }
            .collect { (start, end) ->
                onRangeSelected(
                    start?.let { nockerlClampMillis(it, minMillis, maxMillis) },
                    end?.let { nockerlClampMillis(it, minMillis, maxMillis) },
                )
            }
    }
    DateRangePicker(state = state, modifier = modifier)
}

/**
 * Time picker under the shared contract.
 *
 * @param hour initial hour (0 to 23).
 * @param minute initial minute: snapped to [minuteStep] on the way in AND out.
 * @param onTimeSelected fires with the (step-snapped) hour/minute.
 * @param modifier outer modifier.
 * @param entryMode DIAL/DEFAULT = the M3 clock dial; TEXT = M3 [TimeInput];
 *   WHEEL falls back to the dial (documented).
 * @param minuteStep snap increment (1 = free); M3 has no native stepping, so
 *   the selection snaps on read: the documented clamping behavior.
 * @param is24Hour force 24h; `null` follows the device setting.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NockerlTimePicker(
    hour: Int,
    minute: Int,
    onTimeSelected: (hour: Int, minute: Int) -> Unit,
    modifier: Modifier = Modifier,
    entryMode: NockerlPickerEntryMode = NockerlPickerEntryMode.DEFAULT,
    minuteStep: Int = 1,
    is24Hour: Boolean? = null,
) {
    val state =
        if (is24Hour != null) {
            rememberTimePickerState(
                initialHour = hour,
                initialMinute = nockerlSnapMinutes(minute, minuteStep),
                is24Hour = is24Hour,
            )
        } else {
            rememberTimePickerState(
                initialHour = hour,
                initialMinute = nockerlSnapMinutes(minute, minuteStep),
            )
        }
    LaunchedEffect(state) {
        snapshotFlow { state.hour to state.minute }
            .collect { (h, m) -> onTimeSelected(h, nockerlSnapMinutes(m, minuteStep)) }
    }
    if (entryMode == NockerlPickerEntryMode.TEXT) {
        TimeInput(state = state, modifier = modifier)
    } else {
        TimePicker(state = state, modifier = modifier)
    }
}

/**
 * Snap a minute value to the nearest [step] (the contract's documented clamping:
 * platforms without native stepping honor the contract by snapping). A step of
 * `<= 1` is a no-op; 60 wraps back to 0. Pure + shared with Swift.
 */
fun nockerlSnapMinutes(
    minute: Int,
    step: Int,
): Int {
    if (step <= 1) return minute.coerceIn(0, 59)
    val snapped = (Math.round(minute / step.toFloat()) * step) % 60
    return snapped.coerceIn(0, 59)
}

/** Clamp millis into the contract bounds (either side optional). */
fun nockerlClampMillis(
    value: Long,
    min: Long?,
    max: Long?,
): Long = value.coerceIn(min ?: Long.MIN_VALUE, max ?: Long.MAX_VALUE)

/** [SelectableDates] enforcing the contract's min/max bounds on the calendar. */
@OptIn(ExperimentalMaterial3Api::class)
private fun nockerlSelectableDates(
    min: Long?,
    max: Long?,
): SelectableDates =
    object : SelectableDates {
        override fun isSelectableDate(utcTimeMillis: Long): Boolean =
            (min == null || utcTimeMillis >= min) && (max == null || utcTimeMillis <= max)
    }
